crates/sui-json-rpc/src/name_service.rs

```diff
@@ -154,12 +160,59 @@ impl FromStr for Domain {
             return Err(DomainParseError::LabelsEmpty);
         }
 
+        add_tld_if_missing(&mut labels);
+
         let labels = labels.into_iter().map(ToOwned::to_owned).collect();
 
         Ok(Domain { labels })
     }
 }
 
+/// Parses a separator from the domain string input.
+/// E.g.  `example.sui` -> `.` | example@sui -> `@` | `example*sui` -> `*`
+fn separator(s: &str) -> Result<char, DomainParseError> {
```

Comment:
You can take advantage of `std::str::matches` for this, I think (code offered without testing):

```rust
fn separator(s: &str) -> Result<char, DomainParseError> {
    let mut seps = s.matches(&ACCEPTED_SEPARATORS);
    let Some(sep) = seps.next() else {
        return Ok(ACCEPTED_SEPARATORS[0];
    }

    if !seps.all(|s| s == sep) {
        return Err(DomainParseError::InvalidSeparator)
    }

    Ok(sep)
}
```

Reply:
This won't work here because it changes sep format to `&str` for the ok case. All our separators will be chars!